### ISSUE TITLE
Issuses on spectrograph side

### ISSUE THREAD

**Issue by changgonKim** [sdss/lvmops#5]
Hi, I'm not sure if Evelyn can answer currently, so I will not specify someone.
I made some API based functionality on my lvmscp actor, and this is the example how others can use:
https://github.com/sdss/lvmscp/blob/develop/python/lvmscp/test.py 

However, I made some questions based on the code here: https://github.com/sdss/lvmops/blob/main/lvmops/spec_interface.py 

**- questions on expose function**

1. What situation do we need the shutter (Boolean parameter)?
Since I already added the setting of the shutter for 
each image type(object, arc, dark, bias) on the lvmscp exposure command (in the instrument control software level).
For instance,
object: shutter open
arc: shutter open
dark: shutter closed
bias: shutter closed

Is there any unique situation that the shutter has to be controlled differently with the situations that I mentioned as an example?
If not, it will be better to remove the shutter argument for the expose function.

2. what type of the metadata that wanted to add on the header will be?
I prefer to be python dictionary. However, I can follow any data type that the science team prefer.

3. How about splitting ccdmodes argument to

binning(int)

and the readout mode can be set by an different function.
The possible option is 400, 800, HDR and each stands for

400 is 400MHz readout
800 is 800MHz readout
HDR is 32-bit sampling readout

Please check here:
https://github.com/sdss/lvmscp/blob/develop/python/lvmscp/test.py 

4. There should be a parameter for which spectrograph to have exposure, or if we want all of three spectrographs start the exposure simultaneously, I can make that command on the instrument control software side too.
now we have sp1, sp2, sp3 parameter on the lvmscp side.

Please check the expose command of the lvmscp:
https://github.com/sdss/lvmscp/blob/develop/python/lvmscp/test.py

**Comment by EvelynJ** [comment]
Hi Changgon

Thanks for the questions. I'll try to answer them as best I can (but obviously others who have access can chip in as well).

0. test.py: that looks good, and fits with what we had in mind when we put together the interfaces.

1. As far as I can see, those four exposure settings (object (including acquisition), arc, dark and bias) are the four main types of exposures we will need to take on a regular basis, and I assume even during commissioning we can use those for any non-standard tests and exposures. I don't see any other situation where we would need to control the shutter differently. So if it's easier or makes more sense to code the shutter control elsewhere and remove it from the spec_interface, that's fine.

2. Can you confirm what you mean here? I mean, are you asking what metadata we want to save, or how we want to save it? Either way, I would like to bring this topic up in the next operations meeting to get others opinions (especially those who are more familiar with python) before I give you a solid answer.

For the former question, there is a list of [LVM metadata definitions](https://wiki.sdss.org/display/LVM/LVM+Metadata+Definitions) on the wiki, but it hasn't been touched in over a year and I want to check if there's anything new that needs to be added as the planning for the survey has progressed. 

For the latter question, I think we discussed in August about saving the metadata in headers and fits tables, but I don't remember us coming to a conclusion.
 
3. yes, I agree that splitting the ccdmode argument into binning and readout mode arguments sounds like a good idea. I'm not sure about whether it's better to set the readout mode outside the expose command (as in your test.py), or keep it as an option within the expose command. My worry is that the readout mode might be reset at some point and then we lose some data as it's observed in the wrong mode, so my preference would be to include it as a parameter within the expose command. 

4. At the moment, I see all spectrographs being used simultaneously, with a shutter I believe being used to direct the light from each standard star to the correct spectrograph. There might be cases during troubleshooting or commissioning when we want to expose one spectrograph only, but I will bring this up in the next operations meeting and get back to you with a more solid answer to this question.

**Comment by ndrory** [comment]
Changgon, Evelyn,

I think we need to add 'flat' and 'twi' to the set of {arc, sci, dark, bias} to account for the dome flats and twilight flats. They are no different in shuttering, though.

	N.

> On Jan 17, 2022, at 3:55 AM, changgonKim ***@***.***> wrote:
> 
> 
> Hi, I'm not sure if Evelyn can answer currently, so I will not specify someone.
> I made some API based functionality on my lvmscp actor, and this is the example how others can use:
> https://github.com/sdss/lvmscp/blob/develop/python/lvmscp/test.py
> 
> However, I made some questions based on the code here: https://github.com/sdss/lvmops/blob/main/lvmops/spec_interface.py
> 
> - questions on expose function
> 
> 	• What situation do we need the shutter (Boolean parameter)?
> Since I already added the setting of the shutter for
> each image type(object, arc, dark, bias) on the lvmscp exposure command (in the instrument control software level).
> For instance,
> object: shutter open
> arc: shutter open
> dark: shutter closed
> bias: shutter closed
> Is there any unique situation that the shutter has to be controlled differently with the situations that I mentioned as an example?
> If not, it will be better to remove the shutter argument for the expose function.
> 
> 	• what type of the metadata that wanted to add on the header will be?
> I prefer to be python dictionary. However, I can follow any data type that the science team prefer.
> 
> 	• How about splitting ccdmodes argument to
> 
> binning(int)
> 
> and the readout mode can be set by an different function.
> The possible option is 400, 800, HDR and each stands for
> 
> 400 is 400MHz readout
> 800 is 800MHz readout
> HDR is 32-bit sampling readout
> 
> Please check here:
> https://github.com/sdss/lvmscp/blob/develop/python/lvmscp/test.py
> 
> 	• There should be a parameter for which spectrograph to have exposure, or if we want all of three spectrographs start the exposure simultaneously, I can make that command on the instrument control software side too.
> now we have sp1, sp2, sp3 parameter on the lvmscp side.
> Please check the expose command of the lvmscp:
> https://github.com/sdss/lvmscp/blob/develop/python/lvmscp/test.py
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android. 
> You are receiving this because you are subscribed to this thread.
> 

-- Niv Drory —
McDonald Observatory / Dept. of Astronomy
The University of Texas at Austin
Tel: +1 512 471 6197
http://www.as.utexas.edu/~drory

**Comment by EvelynJ** [comment]
Good point. I had mentally classed them under object, but it would be useful to have them separate for easier identification and for the DRP. 

Changgon- this Thursday will be an operations meeting, so we will discuss all these questions and send you more solid answers by the end of the week.

**Comment by changgonKim** [comment]
Thanks Evelyn, and Niv for the answer.

I think I can still work on several works by your comments.

- Adding 'flat' and 'twi' mode on the exposure, 
- adding readout as a parameter on the exposure command, 
- setting the simultaneous exposure of all three spectrographs as the default

seems to be the work that could be done without discussion. I will work on these, and share you the updates.

For the metadata, I thought we are adding additional data(e.g more than the data on https://wiki.sdss.org/display/LVM/LVM+Metadata+Definitions) that the user wants on the fits header.
For example, someone could want to add the FITS header on the data as

#FITS HEADER
#KEYWORD, VALUE
{"TEST": "1"}

Then in this case, we have to define which data type will be the input.
And the great sample to cover this kind of data type is the python dictionary type as here:
https://realpython.com/python-dicts/ 
So I was just asking if we confirmed that the FITS header data input will be such python dictionary.

I hope to attend the operation meeting and learn how it's going, so please tell me if the meeting time get's fixed.

Thanks!
